File: layers/convlayer.py
from utils.consts    import *
from utils.fmapdesc import  OfmapDesc
from layer          import Layer
from subsamplelayer import SubSampleLayer
import nets.network

##########################################################
class ConvLayer(SubSampleLayer):
    #-----------------------------------------------------------------
    def __init__(self, ntwk, prev_layer, num_ofmaps, stride, kernel):
        assert(isinstance(ntwk, nets.network.Network))
        assert(isinstance(prev_layer, Layer))

        super(ConvLayer, self).__init__(ntwk, prev_layer,
            num_ofmaps=num_ofmaps, stride=stride, kernel=kernel)

    #-----------------------------------------------------------------
    def __str__(self):
        ks = str(self.gKernel())
        ss = str(self.gStride())
        baseLayer = self.gBaseLayerStr()
        if self.gPrevLayer(0) == self.gNetwork().m_PrevLayer:
            return ("Conv " + baseLayer
                    + ", kernel=" + ks + "x" + ks + ", stride=" + ss
                    + self.gStateSizesStr())
        else:
            p = "[" + self.gPrevLayer(0).gNumberStr() + "]"
            return ("Conv" + p + " " + baseLayer
                    + ", kernel=" + ks + "x" + ks + ", stride=" + ss
                    + self.gStateSizesStr())

    def verify(self):
        assert(self.gNumPrevLayers() == 1)
        prevLayer = self.gPrevLayer(0)
        prevMapSize = prevLayer.gOfmapSize()
        assert(prevMapSize/self.gStride() == self.gOfmapSize())
        ##assert(self.gPrevLayer(0).gOfmapSize()/self.gStride() == self.gOfmapSize())


    #-----------------------------------------------------------------
    def gLayerType(self):
        return LAYER_TYPE_CONV

    def gName(self):
        return "Conv"

    #-----------------------------------------------------------------
    def gNumConvWeights(self):
        k = self.gKernel()
        num_ifmaps = gPrevLayer(0).gNumOfmaps()
        num_ofmaps = self.gNumOfmaps()
        return k*k * num_ifmaps * num_ofmaps

    #-----------------------------------------------------------------
    def qPassThrough(self):
        return False

    def gNumberWeights(self):
        assert(self.gNumPrevLayers() == 1)
        prevLayer = self.gPrevLayer(0)
        return prevLayer.gNumOfmaps() * self.gNumOfmaps() * (self.gKernel()**2)

File: compiler/be/layers/src/convlayer.cpp
#include <sstream>

#include "convlayer.hpp"


namespace kcc {

namespace nets {
    class Network;
}

namespace layers {

//--------------------------------------------------------
ConvLayer::ConvLayer(const Params& params, Layer* prev_layer,
        const FmapDesc& fmapDesc,
        const std::tuple<kcc_int32,kcc_int32>& stride, const std::tuple<kcc_int32,kcc_int32>& kernel,
        const std::tuple<kcc_int32,kcc_int32,kcc_int32,kcc_int32>& padding,
        const char* filterFileName, const char* filterTensorFormat)
    : SubSampleLayer(params, prev_layer, fmapDesc, stride, kernel, padding)
{
    m_FilterFileName           = filterFileName;
    m_FilterTensorFormat       = filterTensorFormat;
    m_WeightAddress            = StateBufferAddress_Invalid;
}

//--------------------------------------------------------
std::string
ConvLayer::gString() const
{
    std::stringstream ss;
    const kcc_int32 kh = gKernelHeight();
    const kcc_int32 kw = gKernelWidth();
    ss << gName() << gBaseLayerStr() << ", kernel=" << kh << "x" << kw
       << ",stride=" << gStrideTopBottom() << "/" << gStrideLeftRight() << gStateSizesStr();
    return ss.str();
}

//--------------------------------------------------------
bool
ConvLayer::verify() const
{
    assert(gNumPrevLayers() == 1 && " Conv layer: number of previous layers not 1");
    const bool ok = this->SubClass::verify();
    const Layer* prevLayer = gPrevLayer(0);
    const kcc_int32 prevMapWidth = prevLayer->gOfmapWidth();
    const kcc_int32 prevMapHeight = prevLayer->gOfmapHeight();
    assert(prevMapWidth ==  gStrideLeftRight() * gOfmapWidth() && "Conv layer: Ifmap width != stride * Ofmap width");
    assert(prevMapHeight == gStrideTopBottom() * gOfmapHeight() && "Conv layer: Ifmap height != stride * Ofmap height");
    return ok;
}

//--------------------------------------------------------
kcc_int32
ConvLayer::gNumberWeights() const
{
    assert(gNumPrevLayers() == 1 && "Conv layer: number previous layers not 1");
    const kcc_int32 num_ifmaps = gPrevLayer(0)->gNumOfmaps();
    return num_ifmaps * gNumberWeightsPerPartition();
}

//--------------------------------------------------------
kcc_int64
ConvLayer::gNumberWeightsPerPartition() const
{
    assert(gNumPrevLayers() == 1 && "Conv layer: number previous layers not 1");
    const kcc_int32 kw = gKernelWidth();
    const kcc_int32 kh = gKernelHeight();
    const kcc_int32 num_ofmaps = gNumOfmaps();
    return kw*kh * num_ofmaps;
}



} // namespace layers
} // namespace kcc


